drivers/spi/spi_sifive.c

```diff
@@ -99,50 +100,63 @@ int spi_config(const struct device *dev, uint32_t frequency,
 	return 0;
 }
 
+static ALWAYS_INLINE bool spi_sifive_send_available(const struct device *dev) {
+	return !(sys_read32(SPI_REG(dev, REG_TXDATA)) & SF_TXDATA_FULL);
+}
+
+static ALWAYS_INLINE
 void spi_sifive_send(const struct device *dev, uint16_t frame)
 {
-	while (sys_read32(SPI_REG(dev, REG_TXDATA)) & SF_TXDATA_FULL) {
-	}
-
 	sys_write32((uint32_t) frame, SPI_REG(dev, REG_TXDATA));
 }
 
-uint16_t spi_sifive_recv(const struct device *dev)
+static ALWAYS_INLINE
+bool spi_sifive_recv(const struct device *dev, uint16_t *val)
 {
-	uint32_t val;
+	uint32_t reg = sys_read32(SPI_REG(dev, REG_RXDATA));
 
-	while ((val = sys_read32(SPI_REG(dev, REG_RXDATA))) & SF_RXDATA_EMPTY) {
+	if (reg & SF_RXDATA_EMPTY) {
+		return false;
 	}
-
-	return (uint16_t) val;
+	*val = reg;
+	return true;
 }
 
-void spi_sifive_xfer(const struct device *dev, const bool hw_cs_control)
+static void spi_sifive_xfer(const struct device *dev, const bool hw_cs_control)
 {
 	struct spi_context *ctx = &SPI_DATA(dev)->ctx;
 	uint16_t txd, rxd;
+	int queued_frames = 0;
+
+	while (spi_context_tx_on(ctx) || spi_context_rx_on(ctx) || queued_frames > 0) {
+		bool send = false;
 
-	do {
-		/* Send a frame */
+		/* As long as frames remain to be sent, attempt to queue them on Tx FIFO. If
+		 * the FIFO is full then another attempt will be made next pass. If Rx length
+		 * > Tx length then queue dummy Tx in order to read the requested Rx data.
+		 */
 		if (spi_context_tx_buf_on(ctx)) {
+			send = true;
 			txd = *ctx->tx_buf;
-		} else {
+		} else if (queued_frames == 0) {  /* Implies spi_context_rx_on(). */
+			send = true;
 			txd = 0U;
 		}
 
-		spi_sifive_send(dev, txd);
-
-		spi_context_update_tx(ctx, 1, 1);
-
-		/* Receive a frame */
-		rxd = spi_sifive_recv(dev);
-
-		if (spi_context_rx_buf_on(ctx)) {
-			*ctx->rx_buf = rxd;
+		if (send && spi_sifive_send_available(dev)) {
+			spi_sifive_send(dev, txd);
+			queued_frames++;
+			spi_context_update_tx(ctx, 1, 1);
 		}
 
-		spi_context_update_rx(ctx, 1, 1);
-	} while (spi_context_tx_on(ctx) || spi_context_rx_on(ctx));
+		if (queued_frames > 0 && spi_sifive_recv(dev, &rxd)) {
+			if (spi_context_rx_buf_on(ctx)) {
+				*ctx->rx_buf = rxd;
```

Comment:
Also here explicit conversion should be used (`ctx->rx_buf` points to an `unit8_t`).
And why `rxd` and `txd` are of type `uint16_t`, actually (only 8 lower bits of both are used in fact)?

Reply:
> And why rxd and txd are of type uint16_t, actually (only 8 lower bits of both are used in fact)?

You are right, uint8_t seems appropriate. I added a separate CL to this PR to change these to uint8_t.